docs/book/content/types/objects/error_handling.md

```diff
@@ -169,3 +178,293 @@ The specified structured error information is included in the [`extensions`](htt
+require a new result type which contains the variant `Ok(User)`
+instead of `Ok(Item)`.
+
+Finally it is possible to send a mutation request and handle the
+result. The following example query shows how to handle the result.
+
+```graphql
+{
+  mutation {
+    addItem(name: "", quantity: 0) {
+      ... on Item {
+        name
+      }
+      ... on ValidationErrors {
+        errors {
+          field
+          message
+        }
+      }
+    }
+  }
+}
+```
+
+A useful side effect of this approach is to have partially successful
+paths. Therefore, a if one of multiple paths fails, the result of the
+successful paths are not discarded.
+
+### Example Input Validation (complex)
+
+Instead of using strings do propagated error, it is possible to use
+GraphQL's type system to describe the errors more precisely. For each
+failable input variable a field in a GraphQL object is created. The
+field is set if the validation for that particular field fails. Notice
+that some kind of code generation reduces the unnecessary work. The
+amount of types which are required is significant larger than
+before. Each functions has their custom `ValidationResult` which
+contains only fields provided by the function.
+
+```rust
+#[derive(juniper::GraphQLObject)]
+pub struct Item {
+    name: String,
+    quantity: i32,
+}
+
+#[derive(juniper::GraphQLObject)]
+pub struct ValidationErrorItem {
+    name: Option<String>,
+    quantity: Option<String>,
+}
+
+#[derive(juniper::GraphQLUnion)]
+pub enum GraphQLResult {
+    Ok(Item),
+    Err(ValidationErrorItem),
+}
+
+pub struct Mutation;
+
+#[juniper::graphql_object]
+impl Mutation {
+    fn addItem(&self, name: String, quantity: i32) -> GraphQLResult {
+        let mut error = ValidationErrorItem {
+            name: None,
+            quantity: None,
+        };
+
+        if !(10 <= name.len() && name.len() <= 100) {
+            error.name(Some("between 10 and 100".to_string()));
+        }
+
+        if !(1 <= quantity && quantity <= 10) {
+            error.quantity(Some("between 1 and 10".to_string()));
+        }
+
+        if error.name.is_none() && error.field.is_none() {
+            GraphQLResult::Ok(Item { name, quantity })
+        } else {
+            GraphQLResult::Err(errors)
+        }
+    }
+}
+
+```
+
+```graphql
+{
+  mutation {
+    addItem {
+      ... on Item {
+        name
+      }
+      ... on ValidationErrorsItem {
+        name
+        quantity
+      }
+    }
+  }
+}
+```
+
+Expected errors are handled directly inside the query. Even more, all
+non-critical errors are known in advance.
+
+### Example Input Validation (complex with critical error)
+
+So far only non-critical errors occurred in the examples. Providing
+errors inside the GraphQL schema still allows to return critical
+errors. In the following example, a theoretical database could fail
+and would generate errors. Since it is not common for the database to
+fail, the corresponding error is returned as a critical error.
```

Comment:
```suggestion
Our examples so far have only included non-critical errors. Providing
errors inside the GraphQL schema still allows you to return unexpected critical
errors when they occur.

In the following example, a theoretical database could fail
and would generate errors. Since it is not common for the database to
fail, the corresponding error is returned as a critical error:
```